File: Packages/TeamTasksKit/Sources/TeamTasksContract/Scenarios/MatrixScenarios.swift
import Foundation
import TeamTasksCore

/// The five columns of the permission matrix (docs/CONTRACTS.md §2) around one task.
struct MatrixFixture: Sendable {
    let admin: ContractUser
    let creator: ContractUser
    let assignee: ContractUser
    let other: ContractUser
    let outsider: ContractUser
    let group: GroupFixture
    /// Created by `creator`, assigned to `assignee`.
    let task: TaskItem

    var members: [ContractUser] { [admin, creator, assignee, other] }

    static func make(_ harness: any ContractHarness) async throws -> MatrixFixture {
        let admin = try await harness.user("Admin")
        let creator = try await harness.user("Createur")
        let assignee = try await harness.user("Assigne")
        let other = try await harness.user("Membre")
        let outsider = try await harness.user("Externe")
        let group = try await admin.makeGroup(joinedBy: [creator, assignee, other])
        let task = try await creator.makeTask(in: group.id, "Matrice", assignees: [assignee])
        return MatrixFixture(
            admin: admin, creator: creator, assignee: assignee, other: other, outsider: outsider, group: group, task: task
        )
    }
}

// The §2 matrix, one scenario per row, plus per-group rights and the "creator who left" rule.
extension ContractScenarios {
    static let matrixScenarios: [ContractScenario] = [
        ContractScenario("matrix.seeGroupMembersAndTasks") { harness in
            let fixture = try await MatrixFixture.make(harness)
            for user in fixture.members {
                let summary = try await user.summary(of: fixture.group.id)
                try Verify.that(summary != nil, "\(user.displayName) lists the group")
                let members = try await user.memberIDs(of: fixture.group.id)
                try Verify.equal(members.count, 4, "\(user.displayName) sees every member")
                let tasks = try await user.tasks.tasks(groupId: fixture.group.id, includeOldDone: false)
                try Verify.equal(tasks, [fixture.task], "\(user.displayName) sees the group's tasks")
                let task = try await user.tasks.task(id: fixture.task.id)
                try Verify.equal(task, fixture.task, "\(user.displayName) reads the task")
            }
            let outsiderGroups = try await fixture.outsider.groups.myGroups()
            try Verify.that(outsiderGroups.isEmpty, "a non-member lists no group")
            try await Verify.hidden("members for a non-member") {
                try await fixture.outsider.groups.members(groupId: fixture.group.id)
            }
            try await Verify.hidden("tasks for a non-member") {
                try await fixture.outsider.tasks.tasks(groupId: fixture.group.id, includeOldDone: true)
            }
            try await Verify.fails(with: .notFound, "a task for a non-member") {
                try await fixture.outsider.tasks.task(id: fixture.task.id)
            }
        },

        ContractScenario("matrix.createTask") { harness in
            let fixture = try await MatrixFixture.make(harness)
            for user in fixture.members {
                let draft = TaskDraft(title: Unique.name("Nouvelle"), assigneeIds: [user.id, fixture.admin.id])
                let created = try await Verify.step("\(user.displayName) creates a task") {
                    try await user.tasks.create(groupId: fixture.group.id, draft: draft)
                }
                try Verify.equal(created.createdBy, user.id, "createdBy of a task created by \(user.displayName)")
                let expected = Set([user.id, fixture.admin.id]).sorted { $0.uuidString < $1.uuidString }
                try Verify.equal(created.assigneeIds, expected, "assignees incl. self")
            }
            try await Verify.fails(with: .forbidden, "a non-member creates a task") {
                try await fixture.outsider.tasks.create(groupId: fixture.group.id, draft: TaskDraft(title: "Intrusion"))
            }
            try await Verify.fails(with: .forbidden, "create a task in an unknown group") {
                try await fixture.admin.tasks.create(groupId: UUID(), draft: TaskDraft(title: "Fantôme"))
            }
            let tasks = try await fixture.admin.tasks.tasks(groupId: fixture.group.id, includeOldDone: true)
            try Verify.equal(tasks.count, 5, "four new tasks plus the fixture task")
        },

        ContractScenario("matrix.editTask") { harness in
            let fixture = try await MatrixFixture.make(harness)
            let draft = TaskDraft(title: Unique.name("Modif"), assigneeIds: [fixture.other.id])
            try await Verify.fails(with: .forbidden, "the assignee edits the task") {
                try await fixture.assignee.tasks.update(taskId: fixture.task.id, draft: draft)
            }
            try await Verify.fails(with: .forbidden, "another member edits the task") {
                try await fixture.other.tasks.update(taskId: fixture.task.id, draft: draft)
            }
            try await Verify.fails(with: .notFound, "a non-member edits the task") {
                try await fixture.outsider.tasks.update(taskId: fixture.task.id, draft: draft)
            }
            let unchanged = try await fixture.admin.tasks.task(id: fixture.task.id)
            try Verify.equal(unchanged, fixture.task, "refused edits change nothing")

            let byCreator = try await Verify.step("the creator edits") {
                try await fixture.creator.tasks.update(taskId: fixture.task.id, draft: draft)
            }
            try Verify.equal(byCreator.title, draft.title, "title edited by the creator")
            try Verify.equal(byCreator.assigneeIds, [fixture.other.id], "assignees edited by the creator")
            let adminDraft = TaskDraft(title: Unique.name("Admin"), priority: .high, assigneeIds: [fixture.assignee.id])
            let byAdmin = try await Verify.step("the admin edits") {
                try await fixture.admin.tasks.update(taskId: fixture.task.id, draft: adminDraft)
            }
            try Verify.equal(byAdmin.title, adminDraft.title, "title edited by the admin")
            try Verify.equal(byAdmin.priority, .high, "priority edited by the admin")
            try Verify.equal(byAdmin.createdBy, fixture.creator.id, "createdBy unchanged by an admin edit")
        },

        ContractScenario("matrix.changeStatus") { harness in
            let fixture = try await MatrixFixture.make(harness)
            try await Verify.fails(with: .forbidden, "another member changes the status") {
                try await fixture.other.tasks.setStatus(taskId: fixture.task.id, status: .inProgress)
            }
            try await Verify.fails(with: .notFound, "a non-member changes the status") {
                try await fixture.outsider.tasks.setStatus(taskId: fixture.task.id, status: .inProgress)
            }
            let byAssignee = try await Verify.step("the assignee changes the status") {
                try await fixture.assignee.tasks.setStatus(taskId: fixture.task.id, status: .inProgress)
            }
            try Verify.equal(byAssignee.status, .inProgress, "status set by the assignee")
            try Verify.equal(byAssignee.title, fixture.task.title, "a status change keeps the other fields")
            let byCreator = try await Verify.step("the creator changes the status") {
                try await fixture.creator.tasks.setStatus(taskId: fixture.task.id, status: .done)
            }
            try Verify.equal(byCreator.status, .done, "status set by the creator")
            let byAdmin = try await Verify.step("the admin changes the status") {
                try await fixture.admin.tasks.setStatus(taskId: fixture.task.id, status: .todo)
            }
            try Verify.equal(byAdmin.status, .todo, "status set by the admin")

            _ = try await fixture.creator.reassign(byAdmin, to: [])
            try await Verify.fails(with: .forbidden, "a former assignee changes the status") {
                try await fixture.assignee.tasks.setStatus(taskId: fixture.task.id, status: .done)
            }
        },

        ContractScenario("matrix.deleteTask") { harness in
            let fixture = try await MatrixFixture.make(harness)
            let second = try await fixture.creator.makeTask(in: fixture.group.id, assignees: [fixture.assignee])
            try await Verify.fails(with: .forbidden, "the assignee deletes the task") {
                try await fixture.assignee.tasks.delete(taskId: fixture.task.id)
            }
            try await Verify.fails(with: .forbidden, "another member deletes the task") {
                try await fixture.other.tasks.delete(taskId: fixture.task.id)
            }
            try await Verify.fails(with: .notFound, "a non-member deletes the task") {
                try await fixture.outsider.tasks.delete(taskId: fixture.task.id)
            }
            try await Verify.step("the creator deletes") { try await fixture.creator.tasks.delete(taskId: fixture.task.id) }
            try await Verify.step("the admin deletes") { try await fixture.admin.tasks.delete(taskId: second.id) }
            for user in fixture.members {
                try await Verify.fails(with: .notFound, "\(user.displayName) reads a deleted task") {
                    try await user.tasks.task(id: fixture.task.id)
                }
            }
            try await Verify.fails(with: .notFound, "delete a deleted task") {
                try await fixture.creator.tasks.delete(taskId: fixture.task.id)
            }
            let remaining = try await fixture.admin.tasks.tasks(groupId: fixture.group.id, includeOldDone: true)
            try Verify.that(remaining.isEmpty, "both tasks are gone, got \(remaining)")
        },

        ContractScenario("matrix.groupAdminActions") { harness in
            let fixture = try await MatrixFixture.make(harness)
            let groupId = fixture.group.id
            let nonAdmins = [fixture.creator, fixture.assignee, fixture.other]
            for (index, user) in nonAdmins.enumerated() {
                let target = nonAdmins[(index + 1) % nonAdmins.count]
                try await Verify.fails(with: .forbidden, "\(user.displayName) renames") {
                    try await user.groups.rename(groupId: groupId, name: Unique.name("Pirate"))
                }
                try await Verify.fails(with: .forbidden, "\(user.displayName) deletes the group") {
                    try await user.groups.deleteGroup(groupId: groupId)
                }
                try await Verify.fails(with: .forbidden, "\(user.displayName) reads the invite code") {
                    try await user.groups.inviteCode(groupId: groupId)
                }
                try await Verify.fails(with: .forbidden, "\(user.displayName) regenerates the invite code") {
                    try await user.groups.regenerateInviteCode(groupId: groupId)
                }
                try await Verify.fails(with: .forbidden, "\(user.displayName) changes a role") {
                    try await user.groups.setRole(groupId: groupId, userId: target.id, role: .admin)
                }
                try await Verify.fails(with: .forbidden, "\(user.displayName) removes a member") {
                    try await user.groups.removeMember(groupId: groupId, userId: target.id)
                }
            }
            let outsider = fixture.outsider
            try await Verify.fails(withAnyOf: [.forbidden, .notFound], "a non-member renames") {
                try await outsider.groups.rename(groupId: groupId, name: Unique.name("Pirate"))
            }
            try await Verify.fails(withAnyOf: [.forbidden, .notFound], "a non-member deletes the group") {
                try await outsider.groups.deleteGroup(groupId: groupId)
            }
            try await Verify.fails(with: .forbidden, "a non-member reads the invite code") {
                try await outsider.groups.inviteCode(groupId: groupId)
            }
            try await Verify.fails(with: .forbidden, "a non-member regenerates the invite code") {
                try await outsider.groups.regenerateInviteCode(groupId: groupId)
            }
            try await Verify.fails(with: .forbidden, "a non-member changes a role") {
                try await outsider.groups.setRole(groupId: groupId, userId: fixture.other.id, role: .admin)
            }
            try await Verify.fails(with: .forbidden, "a non-member removes a member") {
                try await outsider.groups.removeMember(groupId: groupId, userId: fixture.other.id)
            }
            let members = try await fixture.admin.memberIDs(of: groupId)
            try Verify.equal(members.count, 4, "refused actions change no membership")

            let admin = fixture.admin
            try await Verify.step("the admin renames, reads and regenerates the code, manages members") {
                _ = try await admin.groups.rename(groupId: groupId, name: Unique.name("Groupe"))
                _ = try await admin.groups.inviteCode(groupId: groupId)
                _ = try await admin.groups.regenerateInviteCode(groupId: groupId)
                try await admin.groups.setRole(groupId: groupId, userId: fixture.other.id, role: .admin)
                try await admin.groups.removeMember(groupId: groupId, userId: fixture.other.id)
                try await admin.groups.deleteGroup(groupId: groupId)
            }
        },

        ContractScenario("matrix.rightsArePerGroup") { harness in
            let alice = try await harness.user("Alice")
            let xavier = try await harness.user("Xavier")
            let aliceGroup = try await alice.makeGroup("Alice", joinedBy: [xavier])
            let xavierGroup = try await xavier.makeGroup("Xavier", joinedBy: [alice])
            let xavierTask = try await xavier.makeTask(in: xavierGroup.id, assignees: [xavier])
            let aliceTask = try await alice.makeTask(in: aliceGroup.id, assignees: [alice])
            let groupId = xavierGroup.id

            // Admin of her own group, Alice is a plain member of Xavier's group.
            try await Verify.fails(with: .forbidden, "rename another group") {
                try await alice.groups.rename(groupId: groupId, name: Unique.name("Pirate"))
            }
            try await Verify.fails(with: .forbidden, "read another group's code") {
                try await alice.groups.inviteCode(groupId: groupId)
            }
            try await Verify.fails(with: .forbidden, "regenerate another group's code") {
                try await alice.groups.regenerateInviteCode(groupId: groupId)
            }
            try await Verify.fails(with: .forbidden, "delete another group") {
                try await alice.groups.deleteGroup(groupId: groupId)
            }
            try await Verify.fails(with: .forbidden, "change a role in another group") {
                try await alice.groups.setRole(groupId: groupId, userId: xavier.id, role: .member)
            }
            try await Verify.fails(with: .forbidden, "remove a member of another group") {
                try await alice.groups.removeMember(groupId: groupId, userId: xavier.id)
            }
            try await Verify.fails(with: .forbidden, "edit a task of another group") {
                try await alice.tasks.update(taskId: xavierTask.id, draft: TaskDraft(title: "Pirate"))
            }
            try await Verify.fails(with: .forbidden, "change the status of a task of another group") {
                try await alice.tasks.setStatus(taskId: xavierTask.id, status: .done)
            }
            try await Verify.fails(with: .forbidden, "delete a task of another group") {
                try await alice.tasks.delete(taskId: xavierTask.id)
            }
            try await Verify.fails(with: .forbidden, "Xavier edits a task of Alice's group") {
                try await xavier.tasks.update(taskId: aliceTask.id, draft: TaskDraft(title: "Pirate"))
            }
            let aliceRole = try await alice.role(in: groupId)
            try Verify.equal(aliceRole, .member, "Alice's role in Xavier's group")
        },

        ContractScenario("matrix.creatorWhoLeftHasNoRights") { harness in
            let alice = try await harness.user("Alice")
            let bob = try await harness.user("Bob")
            let group = try await alice.makeGroup(joinedBy: [bob])
            let task = try await bob.makeTask(in: group.id, assignees: [bob])
            _ = try await Verify.step("the creator edits while a member") {
                try await bob.tasks.update(taskId: task.id, draft: TaskDraft(title: Unique.name("Avant")))
            }
            try await bob.groups.leave(groupId: group.id)

            try await Verify.fails(with: .notFound, "a creator who left reads the task") {
                try await bob.tasks.task(id: task.id)
            }
            try await Verify.fails(with: .notFound, "a creator who left edits the task") {
                try await bob.tasks.update(taskId: task.id, draft: TaskDraft(title: "Pirate"))
            }
            try await Verify.fails(with: .notFound, "a creator who left changes the status") {
                try await bob.tasks.setStatus(taskId: task.id, status: .done)
            }
            try await Verify.fails(with: .notFound, "a creator who left deletes the task") {
                try await bob.tasks.delete(taskId: task.id)
            }
            let kept = try await alice.tasks.task(id: task.id)
            try Verify.equal(kept.createdBy, bob.id, "createdBy kept after the creator left")
            try Verify.equal(kept.assigneeIds, [], "the creator's assignment was deleted")

            let rejoined = try await bob.join(group)
            try Verify.that(!rejoined.alreadyMember, "rejoining is a new membership")
            let role = try await bob.role(in: group.id)
            try Verify.equal(role, .member, "a rejoining member is a plain member")
            let edited = try await Verify.step("the creator edits again once a member") {
                try await bob.tasks.update(taskId: task.id, draft: TaskDraft(title: Unique.name("Après")))
            }
            try Verify.equal(edited.createdBy, bob.id, "createdBy still the creator")
            try await Verify.step("the creator deletes again once a member") { try await bob.tasks.delete(taskId: task.id) }
        },
    ]
}

File: android/contract/src/main/kotlin/io/github/notkanaa/equipe/contract/MatrixScenarios.kt
package io.github.notkanaa.equipe.contract

import io.github.notkanaa.equipe.core.AppError
import io.github.notkanaa.equipe.core.MemberRole
import io.github.notkanaa.equipe.core.TaskDraft
import io.github.notkanaa.equipe.core.TaskItem
import io.github.notkanaa.equipe.core.TaskPriority
import io.github.notkanaa.equipe.core.TaskStatus
import io.github.notkanaa.equipe.core.uuidString
import java.util.UUID

/** The five columns of the permission matrix (docs/CONTRACTS.md §2) around one task. */
internal class MatrixFixture(
    val admin: ContractUser,
    val creator: ContractUser,
    val assignee: ContractUser,
    val other: ContractUser,
    val outsider: ContractUser,
    val group: GroupFixture,
    /** Created by [creator], assigned to [assignee]. */
    val task: TaskItem,
) {
    val members: List<ContractUser> get() = listOf(admin, creator, assignee, other)

    companion object {
        suspend fun make(harness: ContractHarness): MatrixFixture {
            val admin = harness.user("Admin")
            val creator = harness.user("Createur")
            val assignee = harness.user("Assigne")
            val other = harness.user("Membre")
            val outsider = harness.user("Externe")
            val group = admin.makeGroup(joinedBy = listOf(creator, assignee, other))
            val task = creator.makeTask(group.id, "Matrice", assignees = listOf(assignee))
            return MatrixFixture(admin, creator, assignee, other, outsider, group, task)
        }
    }
}

// The §2 matrix, one scenario per row, plus per-group rights and the "creator who left" rule.
internal val matrixScenarios: List<ContractScenario> = listOf(
    ContractScenario("matrix.seeGroupMembersAndTasks") { harness ->
        val fixture = MatrixFixture.make(harness)
        for (user in fixture.members) {
            val summary = user.summary(fixture.group.id)
            Verify.that(summary != null, "${user.displayName} lists the group")
            val members = user.memberIds(fixture.group.id)
            Verify.equal(members.size, 4, "${user.displayName} sees every member")
            val tasks = user.tasks.tasks(fixture.group.id, includeOldDone = false)
            Verify.equal(tasks, listOf(fixture.task), "${user.displayName} sees the group's tasks")
            val task = user.tasks.task(fixture.task.id)
            Verify.equal(task, fixture.task, "${user.displayName} reads the task")
        }
        val outsiderGroups = fixture.outsider.groups.myGroups()
        Verify.that(outsiderGroups.isEmpty(), "a non-member lists no group")
        Verify.hidden("members for a non-member") { fixture.outsider.groups.members(fixture.group.id) }
        Verify.hidden("tasks for a non-member") {
            fixture.outsider.tasks.tasks(fixture.group.id, includeOldDone = true)
        }
        Verify.fails(AppError.NotFound, "a task for a non-member") { fixture.outsider.tasks.task(fixture.task.id) }
    },

    ContractScenario("matrix.createTask") { harness ->
        val fixture = MatrixFixture.make(harness)
        for (user in fixture.members) {
            val draft = TaskDraft(title = Unique.name("Nouvelle"), assigneeIds = setOf(user.id, fixture.admin.id))
            val created = Verify.step("${user.displayName} creates a task") {
                user.tasks.create(fixture.group.id, draft)
            }
            Verify.equal(created.createdBy, user.id, "createdBy of a task created by ${user.displayName}")
            val expected = setOf(user.id, fixture.admin.id).sortedBy { it.uuidString }
            Verify.equal(created.assigneeIds, expected, "assignees incl. self")
        }
        Verify.fails(AppError.Forbidden, "a non-member creates a task") {
            fixture.outsider.tasks.create(fixture.group.id, TaskDraft(title = "Intrusion"))
        }
        Verify.fails(AppError.Forbidden, "create a task in an unknown group") {
            fixture.admin.tasks.create(UUID.randomUUID(), TaskDraft(title = "Fantôme"))
        }
        val tasks = fixture.admin.tasks.tasks(fixture.group.id, includeOldDone = true)
        Verify.equal(tasks.size, 5, "four new tasks plus the fixture task")
    },

    ContractScenario("matrix.editTask") { harness ->
        val fixture = MatrixFixture.make(harness)
        val draft = TaskDraft(title = Unique.name("Modif"), assigneeIds = setOf(fixture.other.id))
        Verify.fails(AppError.Forbidden, "the assignee edits the task") {
            fixture.assignee.tasks.update(fixture.task.id, draft)
        }
        Verify.fails(AppError.Forbidden, "another member edits the task") {
            fixture.other.tasks.update(fixture.task.id, draft)
        }
        Verify.fails(AppError.NotFound, "a non-member edits the task") {
            fixture.outsider.tasks.update(fixture.task.id, draft)
        }
        val unchanged = fixture.admin.tasks.task(fixture.task.id)
        Verify.equal(unchanged, fixture.task, "refused edits change nothing")

        val byCreator = Verify.step("the creator edits") { fixture.creator.tasks.update(fixture.task.id, draft) }
        Verify.equal(byCreator.title, draft.title, "title edited by the creator")
        Verify.equal(byCreator.assigneeIds, listOf(fixture.other.id), "assignees edited by the creator")
        val adminDraft = TaskDraft(
            title = Unique.name("Admin"), priority = TaskPriority.HIGH, assigneeIds = setOf(fixture.assignee.id),
        )
        val byAdmin = Verify.step("the admin edits") { fixture.admin.tasks.update(fixture.task.id, adminDraft) }
        Verify.equal(byAdmin.title, adminDraft.title, "title edited by the admin")
        Verify.equal(byAdmin.priority, TaskPriority.HIGH, "priority edited by the admin")
        Verify.equal(byAdmin.createdBy, fixture.creator.id, "createdBy unchanged by an admin edit")
    },

    ContractScenario("matrix.changeStatus") { harness ->
        val fixture = MatrixFixture.make(harness)
        Verify.fails(AppError.Forbidden, "another member changes the status") {
            fixture.other.tasks.setStatus(fixture.task.id, TaskStatus.IN_PROGRESS)
        }
        Verify.fails(AppError.NotFound, "a non-member changes the status") {
            fixture.outsider.tasks.setStatus(fixture.task.id, TaskStatus.IN_PROGRESS)
        }
        val byAssignee = Verify.step("the assignee changes the status") {
            fixture.assignee.tasks.setStatus(fixture.task.id, TaskStatus.IN_PROGRESS)
        }
        Verify.equal(byAssignee.status, TaskStatus.IN_PROGRESS, "status set by the assignee")
        Verify.equal(byAssignee.title, fixture.task.title, "a status change keeps the other fields")
        val byCreator = Verify.step("the creator changes the status") {
            fixture.creator.tasks.setStatus(fixture.task.id, TaskStatus.DONE)
        }
        Verify.equal(byCreator.status, TaskStatus.DONE, "status set by the creator")
        val byAdmin = Verify.step("the admin changes the status") {
            fixture.admin.tasks.setStatus(fixture.task.id, TaskStatus.TODO)
        }
        Verify.equal(byAdmin.status, TaskStatus.TODO, "status set by the admin")

        fixture.creator.reassign(byAdmin, emptyList())
        Verify.fails(AppError.Forbidden, "a former assignee changes the status") {
            fixture.assignee.tasks.setStatus(fixture.task.id, TaskStatus.DONE)
        }
    },

    ContractScenario("matrix.deleteTask") { harness ->
        val fixture = MatrixFixture.make(harness)
        val second = fixture.creator.makeTask(fixture.group.id, assignees = listOf(fixture.assignee))
        Verify.fails(AppError.Forbidden, "the assignee deletes the task") {
            fixture.assignee.tasks.delete(fixture.task.id)
        }
        Verify.fails(AppError.Forbidden, "another member deletes the task") {
            fixture.other.tasks.delete(fixture.task.id)
        }
        Verify.fails(AppError.NotFound, "a non-member deletes the task") {
            fixture.outsider.tasks.delete(fixture.task.id)
        }
        Verify.step("the creator deletes") { fixture.creator.tasks.delete(fixture.task.id) }
        Verify.step("the admin deletes") { fixture.admin.tasks.delete(second.id) }
        for (user in fixture.members) {
            Verify.fails(AppError.NotFound, "${user.displayName} reads a deleted task") {
                user.tasks.task(fixture.task.id)
            }
        }
        Verify.fails(AppError.NotFound, "delete a deleted task") { fixture.creator.tasks.delete(fixture.task.id) }
        val remaining = fixture.admin.tasks.tasks(fixture.group.id, includeOldDone = true)
        Verify.that(remaining.isEmpty(), "both tasks are gone, got $remaining")
    },

    ContractScenario("matrix.groupAdminActions") { harness ->
        val fixture = MatrixFixture.make(harness)
        val groupId = fixture.group.id
        val nonAdmins = listOf(fixture.creator, fixture.assignee, fixture.other)
        for ((index, user) in nonAdmins.withIndex()) {
            val target = nonAdmins[(index + 1) % nonAdmins.size]
            Verify.fails(AppError.Forbidden, "${user.displayName} renames") {
                user.groups.rename(groupId, Unique.name("Pirate"))
            }
            Verify.fails(AppError.Forbidden, "${user.displayName} deletes the group") {
                user.groups.deleteGroup(groupId)
            }
            Verify.fails(AppError.Forbidden, "${user.displayName} reads the invite code") {
                user.groups.inviteCode(groupId)
            }
            Verify.fails(AppError.Forbidden, "${user.displayName} regenerates the invite code") {
                user.groups.regenerateInviteCode(groupId)
            }
            Verify.fails(AppError.Forbidden, "${user.displayName} changes a role") {
                user.groups.setRole(groupId, target.id, MemberRole.ADMIN)
            }
            Verify.fails(AppError.Forbidden, "${user.displayName} removes a member") {
                user.groups.removeMember(groupId, target.id)
            }
        }
        val outsider = fixture.outsider
        Verify.fails(AppError.Forbidden, "a non-member renames") {
            outsider.groups.rename(groupId, Unique.name("Pirate"))
        }
        Verify.fails(AppError.Forbidden, "a non-member deletes the group") { outsider.groups.deleteGroup(groupId) }
        Verify.fails(AppError.Forbidden, "a non-member reads the invite code") { outsider.groups.inviteCode(groupId) }
        Verify.fails(AppError.Forbidden, "a non-member regenerates the invite code") {
            outsider.groups.regenerateInviteCode(groupId)
        }
        Verify.fails(AppError.Forbidden, "a non-member changes a role") {
            outsider.groups.setRole(groupId, fixture.other.id, MemberRole.ADMIN)
        }
        Verify.fails(AppError.Forbidden, "a non-member removes a member") {
            outsider.groups.removeMember(groupId, fixture.other.id)
        }
        val members = fixture.admin.memberIds(groupId)
        Verify.equal(members.size, 4, "refused actions change no membership")

        // Lead decision: on an unknown group these admin RPCs answer forbidden (not "not found").
        val admin = fixture.admin
        val unknown = UUID.randomUUID()
        Verify.fails(AppError.Forbidden, "regenerate the invite code of an unknown group") {
            admin.groups.regenerateInviteCode(unknown)
        }
        Verify.fails(AppError.Forbidden, "change a role in an unknown group") {
            admin.groups.setRole(unknown, fixture.other.id, MemberRole.ADMIN)
        }
        Verify.fails(AppError.Forbidden, "remove a member of an unknown group") {
            admin.groups.removeMember(unknown, fixture.other.id)
        }

        Verify.step("the admin renames, reads and regenerates the code, manages members") {
            admin.groups.rename(groupId, Unique.name("Groupe"))
            admin.groups.inviteCode(groupId)
            admin.groups.regenerateInviteCode(groupId)
            admin.groups.setRole(groupId, fixture.other.id, MemberRole.ADMIN)
            admin.groups.removeMember(groupId, fixture.other.id)
            admin.groups.deleteGroup(groupId)
        }
    },

    ContractScenario("matrix.rightsArePerGroup") { harness ->
        val alice = harness.user("Alice")
        val xavier = harness.user("Xavier")
        val aliceGroup = alice.makeGroup("Alice", joinedBy = listOf(xavier))
        val xavierGroup = xavier.makeGroup("Xavier", joinedBy = listOf(alice))
        val xavierTask = xavier.makeTask(xavierGroup.id, assignees = listOf(xavier))
        val aliceTask = alice.makeTask(aliceGroup.id, assignees = listOf(alice))
        val groupId = xavierGroup.id

        // Admin of her own group, Alice is a plain member of Xavier's group.
        Verify.fails(AppError.Forbidden, "rename another group") {
            alice.groups.rename(groupId, Unique.name("Pirate"))
        }
        Verify.fails(AppError.Forbidden, "read another group's code") { alice.groups.inviteCode(groupId) }
        Verify.fails(AppError.Forbidden, "regenerate another group's code") {
            alice.groups.regenerateInviteCode(groupId)
        }
        Verify.fails(AppError.Forbidden, "delete another group") { alice.groups.deleteGroup(groupId) }
        Verify.fails(AppError.Forbidden, "change a role in another group") {
            alice.groups.setRole(groupId, xavier.id, MemberRole.MEMBER)
        }
        Verify.fails(AppError.Forbidden, "remove a member of another group") {
            alice.groups.removeMember(groupId, xavier.id)
        }
        Verify.fails(AppError.Forbidden, "edit a task of another group") {
            alice.tasks.update(xavierTask.id, TaskDraft(title = "Pirate"))
        }
        Verify.fails(AppError.Forbidden, "change the status of a task of another group") {
            alice.tasks.setStatus(xavierTask.id, TaskStatus.DONE)
        }
        Verify.fails(AppError.Forbidden, "delete a task of another group") { alice.tasks.delete(xavierTask.id) }
        Verify.fails(AppError.Forbidden, "Xavier edits a task of Alice's group") {
            xavier.tasks.update(aliceTask.id, TaskDraft(title = "Pirate"))
        }
        val aliceRole = alice.role(groupId)
        Verify.equal(aliceRole, MemberRole.MEMBER, "Alice's role in Xavier's group")
    },

    ContractScenario("matrix.creatorWhoLeftHasNoRights") { harness ->
        val alice = harness.user("Alice")
        val bob = harness.user("Bob")
        val group = alice.makeGroup(joinedBy = listOf(bob))
        val task = bob.makeTask(group.id, assignees = listOf(bob))
        Verify.step("the creator edits while a member") {
            bob.tasks.update(task.id, TaskDraft(title = Unique.name("Avant")))
        }
        bob.groups.leave(group.id)

        Verify.fails(AppError.NotFound, "a creator who left reads the task") { bob.tasks.task(task.id) }
        Verify.fails(AppError.NotFound, "a creator who left edits the task") {
            bob.tasks.update(task.id, TaskDraft(title = "Pirate"))
        }
        Verify.fails(AppError.NotFound, "a creator who left changes the status") {
            bob.tasks.setStatus(task.id, TaskStatus.DONE)
        }
        Verify.fails(AppError.NotFound, "a creator who left deletes the task") { bob.tasks.delete(task.id) }
        val kept = alice.tasks.task(task.id)
        Verify.equal(kept.createdBy, bob.id, "createdBy kept after the creator left")
        Verify.equal(kept.assigneeIds, emptyList<UUID>(), "the creator's assignment was deleted")

        val rejoined = bob.join(group)
        Verify.that(!rejoined.alreadyMember, "rejoining is a new membership")
        val role = bob.role(group.id)
        Verify.equal(role, MemberRole.MEMBER, "a rejoining member is a plain member")
        val edited = Verify.step("the creator edits again once a member") {
            bob.tasks.update(task.id, TaskDraft(title = Unique.name("Après")))
        }
        Verify.equal(edited.createdBy, bob.id, "createdBy still the creator")
        Verify.step("the creator deletes again once a member") { bob.tasks.delete(task.id) }
    },
)
